Library/OcDeviceMiscLib/ForgeUefi.c

```diff
@@ -93,14 +115,34 @@ OcForgeUefiSupport (
     return EFI_INVALID_PARAMETER;
   }
 
-  NewBS = AllocateZeroPool (sizeof (EFI_BOOT_SERVICES));
-  if (NewBS == NULL) {
-    DEBUG ((DEBUG_INFO, "OCDM: Failed to allocate BS copy\n"));
-    return EFI_OUT_OF_RESOURCES;
+  if (Trash) {
+    Signature = *(UINT64 *)(&gBS->CreateEventEx);
+    if (Signature != DXE_SERVICES_SIGNATURE) {
```

Comment:
Does this ever happen?

Reply:
No. I thought it was safer to check that the memory is safe to trash. See comment at start of this function.

Reply:
i.e. it would only ever happen if injected/used on some other firmware, but I guess _that_ could definitely happen.

Reply:
Just I would rather abort and try to reallocate. Reallocting boot services will likely cause hard to debug hibernation or boot issues, so better not to put the mines for us here.

Reply:
Hang on though - if "reallocting boot services will likely cause hard to debug hibernation or boot issues", then my best bet is to abort the forge attempt and do nothing (won't forge UEFI, driver probably won't work, but no such hard to debug problems). Which is what the code does, no? (What have I missed?!)

Reply:
Yes, abort and do nothing. I only wanted to remove the reallocation part, it will not do us any good I fear.

Reply:
Disregard this comment. Realised that I was wrong with the suggestions above. The issue is that the driver in DXE mode cannot reallocate the BS table and thus needs to patch it.